src/workerd/api/sockets.c++

```diff
@@ -346,9 +352,28 @@ jsg::Ref<Socket> Socket::startTls(jsg::Lock& js, jsg::Optional<TlsOptions> tlsOp
     // that's the case.
     JSG_REQUIRE(
         *tlsStarter != kj::none, TypeError, "The request has finished before startTls completed.");
-    auto secureStream = KJ_ASSERT_NONNULL(*tlsStarter)(acceptedHostname)
+
+    auto& context = IoContext::current();
+    auto tlsStarterResult =
+        context.awaitIo(js, KJ_ASSERT_NONNULL(*tlsStarter)(acceptedHostname))
+            .then(js,
+                [openedResolverResolve = kj::mv(openedResolverResolve),
+                    remoteAddress = kj::mv(remoteAddress)](jsg::Lock& js) mutable {
+      openedResolverResolve.resolve(js,
+          SocketInfo{
+            .remoteAddress = kj::mv(remoteAddress),
+            .localAddress = kj::none,
+          });
+    },
+                [openedResolverReject = kj::mv(openedResolverReject)](
+                    jsg::Lock& js, jsg::Value&& error) mutable {
+      openedResolverReject.reject(js, jsg::JsValue(error.getHandle(js)));
+    });
+
+    auto secureStream = context.awaitJs(js, kj::mv(tlsStarterResult))
```

Comment:
This awaitJs() is waiting on the result of an awaitIo(), which creates a weird need to jump from KJ event loop to JS microtask loop and back to KJ event loop. Could this be avoided by instead forking the TLS starter promise, passing one branch to awaitIo and waiting directly on the other branch here?